compiler/rustc_trait_selection/src/opaque_types.rs

```diff
@@ -1139,18 +1161,22 @@ impl<'a, 'tcx> Instantiator<'a, 'tcx> {
         // Foo, impl Bar)`.
         let definition_span = self.value_span;
 
-        self.opaque_types.insert(
-            def_id,
-            OpaqueTypeDecl {
-                opaque_type: ty,
-                substs,
-                definition_span,
-                concrete_ty: ty_var,
-                has_required_region_bounds: !required_region_bounds.is_empty(),
-                origin,
-            },
-        );
-        debug!("instantiate_opaque_types: ty_var={:?}", ty_var);
+        // We only keep the first concrete type var, as we will already error
+        // out if there are multiple due to the conflicting obligations
+        if !self.opaque_types.contains_key(&def_id) {
```

Comment:
So, I'm guessing we're not returning early above so that the predicates can be checked for *both* uses right? I guess, I would expect that we would first find the defining use (or error out if there isn't one or multiple), then later come back and check bounds if there is one and only one.

Reply:
> So, I'm guessing we're not returning early above so that the predicates can be checked for both uses right?

yes

> I would expect that we would first find the defining use (or error out if there isn't one or multiple), then later come back and check bounds if there is one and only one.

This is not the code that looks for defining uses, this is the code that generates type variables for inference to figure out the concrete type. While the diagnostic talks about defining uses, this is just to have the diagnostic be similar to the diagnostic we get for multiple conflicting defining uses that are in different functions. Since we are within a single function, we just "assume" these are defining uses (and will always be right about that assumption).